### DISCUSSION TITLE
Can i use hardware video encoding with encodarr runner ?

### DISCUSSION THREAD

**Discussion by tr1p0p** [BrenekH/encodarr#98] [Q&A] [answered]
Hi again,

I wonder how and where i can see the ffmpeg command encodarr runner do when it receive a file from the controller. I have a AMD GPU and i would like to know if ffmpeg take advantages of my raedon card to help for the encoding.

Where can we change the type on encoding ? I want to set ffmpeg with `hevc_amf` 

Thanks.

**Comment by BrenekH** [comment, marked as answer]
Thanks for reaching out!

I don't have the hardware to test fully, which is why I've never worried about it, but I've added the ability to use a hardware codec in [Version 0.2.2](https://github.com/BrenekH/encodarr/releases/0.2.2). You will still need to set Target Video Codec to the correct value because it's still used when decided whether or not to run a job.

**Reply by tr1p0p** [reply]
Thanks a lot ! So i put the terms like that (was vaapi i was looking actually : `hevc_vaapi` but i don't see much activity on my GPU after, did i just have to just add "vaapi" without the hevc terms ?

**Reply by BrenekH** [reply]
No, it needs to be `hevc_vaapi` but it looks like the device may need to be set([source](https://stackoverflow.com/a/28209208/13298346)). Since I don't have hardware access I'm afraid won't be of much help. If you could find a `ffmpeg` command that does what you want and send it to me, maybe I can figure out how to fit it into Encodarr.

You should also enable trace logging for the Runner so that it records the FFmpeg output. That could help determine if it's actually working or not.

**Reply by tr1p0p** [reply]
Ok so from [this page](https://trac.ffmpeg.org/wiki/Hardware/VAAPI) i get this line that encode properly on my linux machine :

`ffmpeg -hwaccel vaapi -hwaccel_output_format vaapi -hwaccel_device /dev/dri/renderD128 -i input.mp4 -c:v h264_vaapi output.mp4`

will test some stuffs with -trace on runner.

**Reply by tr1p0p** [reply]
So, it's seem the more easy way to put it in a ffmpeg command would be :

`ffmpeg [set device for vaapi] [video file] [options] [encoder] output.mp4`

Maybe like, a hidden value that appear when we activate hardware encoding that can serve to set the device ? (it's very often /dev/dri/renderD128, at least on linux)

**Reply by BrenekH** [reply]
The problem with setting the device on the Controller is that not all Runners can be guaranteed to have the same hardware at the same path (and Windows vs Linux would be different as well). I suppose I can add a device field and a warning with a link to a page with more details, but it's not a final solution.

Hopefully, this problem can be resolved with the plugin system and another Runner setting. Although, I don't know when that will all be ready, probably won't be for a few months at the least.

**Reply by BrenekH** [reply]
Ok, I've got a first draft of this ready. Can you test it to make sure everything works as expected?

[Controller](https://github.com/BrenekH/encodarr/actions/runs/1076269973)
[Runner](https://github.com/BrenekH/encodarr/actions/runs/1076269974)

**Reply by tr1p0p** [reply]
Ok, so with `-trace` log i don't see no mention of vaapi device or anything, i can not tell you if it's working or not.

**Reply by BrenekH** [reply]
To enable trace logging you need to use `--log-level trace`. Then the logs will outputted to the console and to `~/encodarr/runner/<current time>/config/runner.log`.

**Reply by tr1p0p** [reply]
Well, this is the logs, maybe you will see : https://paste.ubuntu.com/p/6fB6NjdPgM/

**Reply by BrenekH** [reply]
Looks like everything is working to me. This is with hardware encoding turned on, correct?

**Reply by tr1p0p** [reply]
Yeah this is my settings.
![Screenshot 2021-07-29 at 01-14-10 Encodarr](https://user-images.githubusercontent.com/37407065/127408038-5e3aaec8-5792-44e3-b507-261239257deb.png)
Well if it's look good for you it's true it's take less cpu and i having a little more fps. Thanks for all the help with this

**Reply by BrenekH** [reply]
Sweet! Glad it's working.

**Reply by tr1p0p** [reply]
I wanna correct a thing i said : `ffmpeg -hwaccel vaapi -hwaccel_device` doesn't encode at all, it mean for decoding only (https://trac.ffmpeg.org/wiki/Hardware/VAAPI)

So the right way of launching a encode should begin with : `ffmpeg -vaapi_device /dev/dri/renderD128 -i [video file] -c:v hevc_vaapi [output video]` 

Just for saying.

EDIT : Also, i try a lot the hardware option lately : i confirm it do nothing in the fps, actually it *reduce* it, i lose like 10 fps when i activate the hardware encoding. So i try to understand why
